crates/contract/src/instance.rs

```diff
@@ -16,79 +17,70 @@ pub struct ContractInstance<P> {
 
 impl<P> ContractInstance<P> {
     /// Creates a new contract from the provided address, provider, and interface.
+    #[inline]
     pub const fn new(address: Address, provider: P, interface: Interface) -> Self {
         Self { address, provider, interface }
     }
 
-    /// Returns the contract's address.
-    pub const fn address(&self) -> Address {
-        self.address
+    /// Returns a reference to the contract's address.
+    #[inline]
+    pub const fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Sets the contract's address.
+    #[inline]
+    pub fn set_address(&mut self, address: Address) {
+        self.address = address;
+    }
+
+    /// Returns a new contract instance at `address`.
+    #[inline]
+    pub fn at(mut self, address: Address) -> ContractInstance<P> {
+        self.set_address(address);
+        self
     }
 
     /// Returns a reference to the contract's ABI.
+    #[inline]
     pub const fn abi(&self) -> &JsonAbi {
         self.interface.abi()
     }
 
     /// Returns a reference to the contract's provider.
-    pub const fn provider_ref(&self) -> &P {
+    #[inline]
+    pub const fn provider(&self) -> &P {
         &self.provider
     }
 }
 
-impl<P> ContractInstance<P>
-where
-    P: Clone,
-{
-    /// Returns a clone of the contract's provider.
-    pub fn provider(&self) -> P {
-        self.provider.clone()
-    }
-
-    /// Returns a new contract instance at `address`.
-    ///
-    /// Clones `self` internally
-    #[must_use]
-    pub fn at(&self, address: Address) -> ContractInstance<P> {
-        let mut this = self.clone();
-        this.address = address;
-        this
+impl<P: Clone> ContractInstance<&P> {
+    /// Clones the provider and returns a new contract instance with the cloned provider.
+    #[inline]
+    pub fn with_cloned_provider(self, provider: P) -> ContractInstance<P> {
+        ContractInstance { address: self.address, provider, interface: self.interface }
```

Comment:
This does not clone the provider but swaps it from a ref to owned

looks like this doesn't need the P: Clone bound and the function could accept any Provider type?